### PR TITLE
Ensure we not log missleading errors if the promise was already faile…

### PR DESCRIPTION
**Pull request by normanmaurer**
…d due errors

Motivation:

In DefaultHttp2ConnectionEncoder we fail the promise in in the FlowControlledData.error(...) method but also add it the CoalescingBufferQueue. Which can lead to have the promise failed by error(...) before it can be failed in CoalescingBufferQueue.

This can lead to confusing and missleading errors in the log like:
    2016-08-12 09:47:43,716       nettyIoExecutorGroup-1-9 [WARN ] PromiseNotifier                - Failed to mark a promise as failure because it's done already: DefaultChannelPromise@374225e0(failure: javax.net.ssl.SSLException: SSLEngine closed already)
    javax.net.ssl.SSLException: SSLEngine closed already
        at io.netty.handler.ssl.SslHandler.wrap(...)(Unknown Source) ~[netty-all-4.1.5.Final-SNAPSHOT.jar:?]

Modifications:

Ensure we not log an error if we fail to notify the future as part of the CoalescingBufferQueue.

Result:

No more missleading logs.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@Scottmitch @nmittler PTAL

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java [hunk 1]
```diff
@@ -341,7 +342,8 @@ public FlowControlledData(Http2Stream stream, ByteBuf buf, int padding, boolean
                                    ChannelPromise promise) {
             super(stream, padding, endOfStream, promise);
             queue = new CoalescingBufferQueue(promise.channel());
-            queue.add(buf, promise);
+            // Using false for the notifier as the may already fail the promise in error(...)
+            queue.add(buf, promise.isVoid() ? null : new ChannelPromiseNotifier(false, promise));
```
Comment (netkins) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java]: MAJOR Ternary operators that can be simplified with || or && [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3ASimplifiedTernary)

Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java]: Why did you add a notifier and not just use the original promise?

Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java]: @nmittler because if a promise was added we will call `new ChannelPromiseNotifier(promise)` internally (without the `false`). But as we also notify the promise outside of the queue we need to ensure we not log if we already completed it.

Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java]: Ah I see.  Can you point me to where the notifier would be created if we don't create it here?

Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java]: https://github.com/netty/netty/blob/4.1/transport/src/main/java/io/netty/channel/CoalescingBufferQueue.java#L69

Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java]: can we instead just let the `queue` be responsible for failing the promise (and remove the `promise.tryFailure(cause);` at the end of `error`)?

Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java]: @Scottmitch good point. This should work as well. Let me update.

**Comment by normanmaurer** [comment]
@Scottmitch PTAL again

**Comment by Scottmitch** [comment]
lgtm .. @nmittler wdyt

**Comment by nmittler** [comment]
LGTM

oof ... we must be getting a lot of log spam for this. Ship it.

**Comment by normanmaurer** [comment]
@nmittler yep saw a lot of these ;)

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (5fd239c)
